File: Javascript/OnesAndZeroes.js
// Leetcode Problem - 474

var findMaxForm = function (strs, m, n) {
  const dp = {};

  function dfs(i, m, n) {
    if (i === strs.length) {
      return 0;
    }

    if (dp[i] && dp[i][m] && dp[i][m][n]) {
      return dp[i][m][n];
    }

    const [mCount, nCount] = countZeroesOnes(strs[i]);
    dp[i] = dp[i] || {};
    dp[i][m] = dp[i][m] || {};
    dp[i][m][n] = dfs(i + 1, m, n);

    if (mCount <= m && nCount <= n) {
      dp[i][m][n] = Math.max(
        dp[i][m][n],
        1 + dfs(i + 1, m - mCount, n - nCount)
      );
    }

    return dp[i][m][n];
  }

  function countZeroesOnes(str) {
    let mCount = 0;
    let nCount = 0;

    for (const char of str) {
      if (char === "0") {
        mCount++;
      } else if (char === "1") {
        nCount++;
      }
    }

    return [mCount, nCount];
  }

  return dfs(0, m, n);
};

File: Python/OnesAndZeroes.py
## Leetcode Problem - 474
def findMaxForm(strs,m,n):

    dp = {}

    def dfs(i,m,n):
        if i==len(strs):
            return 0
        
        if (i,m,n) in dp:
            return dp[(i,m,n)]
        
        mCount,nCount = strs[i].count("0"),strs[i].count("1")
        dp[(i,m,n)] = dfs(i+1,m,n)
        
        if mCount<=m and nCount<=n:
            dp[(i,m,n)] = max(dp[(i,m,n)] , 1+dfs(i+1,m-mCount,n-nCount))

        return dp[(i,m,n)]
    
    return dfs(0,m,n)

strs = ["10","0001","111001","1","0"]
m = 5
n = 3

print(findMaxForm(strs,m,n))